### ISSUE TITLE
Intermittently, only latest journal entry is sent to actor as SnapshotOffer

### ISSUE THREAD

**Issue by boekabart** [akkadotnet/Akka.Persistence.PostgreSql#66]
An application has a single Typed Persistent Actor.
This actor is created (and therefore, recovered) right at the start of the application.
The actor does not (yet) implement snapshotting, at all.

In (Pre)Production (never in dev environment), in about 50% of the startups, recovery goes completely wrong.
Instead of receiving the thousands of journal entries, it receives ONLY the latestmost journal entry, but instead of just receiving it, it gets it presented as the Snapshot of a SnapshotOffer.

Most relevant piece of logging:

```
2019-08-13 12:27:54,032Z DEBUG [5  ] PostgreSqlSnapshotStore                  Started (Akka.Persistence.PostgreSql.Snapshot.PostgreSqlSnapshotStore)
2019-08-13 12:27:54,047Z DEBUG [8  ] PostgreSqlJournal                        Started (Akka.Persistence.PostgreSql.Journal.PostgreSqlJournal)
2019-08-13 12:27:54,625Z DEBUG [8  ] RuntimeActor                             Recovering an Snapshot 
snapshotOffer.ToString: SnapshotOffer<meta: SnapshotMetadata<pid: RuntimeActor, seqNr: 66298, timestamp: 2019/08/09>, snapshot: de7a22d1-bd93-4fef-81c3-00878a9f3517 en-queued at 2019-08-09T15:55:33.9508626+02:00>
snapshotOffer.SnapShot.GetType: Divv.IxIx.TitleCatalogizer.Actors.Runtime.RuntimeActor+EnqueuedWork
snapshotOffer.SnapShot.ToString de7a22d1-bd93-4fef-81c3-00878a9f3517 en-queued at 2019-08-09T15:55:33.9508626+02:00
2019-08-13 12:27:54,750Z DEBUG [5  ] ReplayFilter                             Started (Akka.Persistence.Journal.ReplayFilter)
2019-08-13 12:27:54,766Z DEBUG [5  ] PostgreSqlJournal
```

Using a client to inspect the 3 tables in postgres, only `event_journal` is populated, `metadata` and `snapshot_store` are empty, 0 rows.

Config:
```
akka : {
  log-config-on-start : on
  log-dead-letters : on
  stdout-loglevel : DEBUG
  loglevel : DEBUG
  loggers : ["Akka.Logger.log4net.Log4NetLogger, Akka.Logger.log4net","Akka.Event.StandardOutLogger, Akka"]
  suppress-json-serializer-warning : on
  actor : {
    debug : {
      receive : on
      autoreceive : on
      lifecycle : on
      event-stream : on
      unhandled : on
    }
  }
  persistence : {
    journal : {
      plugin : akka.persistence.journal.postgresql
      postgresql : {
        class : "Akka.Persistence.PostgreSql.Journal.PostgreSqlJournal, Akka.Persistence.PostgreSql"
        plugin-dispatcher : akka.actor.default-dispatcher
        connection-string : "Server=divv-1-pgsql-l.xixi.com;Port=5432;User Id=RtiTitleCatalogizer;Password=****;Database=RtiTitleCatalogizer;SslMode=Disable;"
        schema-name : public
        connection-timeout : 30s
        auto-initialize : on
        timestamp-provider : "Akka.Persistence.Sql.Common.Journal.DefaultTimestampProvider, Akka.Persistence.Sql.Common"
        metadata-table-name : metadata
        stored-as : BYTEA
      }
    }
    snapshot-store : {
      plugin : akka.persistence.snapshot-store.postgresql
      postgresql : {
        class : "Akka.Persistence.PostgreSql.Snapshot.PostgreSqlSnapshotStore, Akka.Persistence.PostgreSql"
        plugin-dispatcher : akka.actor.default-dispatcher
        connection-string : "Server=divv-1-pgsql-l.xixi.com;Port=5432;User Id=RtiTitleCatalogizer;Password=****;Database=RtiTitleCatalogizer;SslMode=Disable;"
        schema-name : public
        connection-timeout : 30s
        timestamp-provider : "Akka.Persistence.Sql.Common.Journal.DefaultTimestampProvider, Akka.Persistence.Sql.Common"
        auto-initialize : on
        stored-as : BYTEA
      }
    }
  }
}
```

When switching to Sqlite, the actor always correctly recovers.

**Comment by boekabart** [comment]
Could this be slightly related to #65 ?

**Comment by boekabart** [comment]
For clarity: in the other ~ 50% of cases, recovery proceeds as normal: no `SnapshotOffer` is received by the actor, and all Journal entries are recovered.

**Comment by boekabart** [comment]
I found a clear correlation between smth in the logging, and the recovery leading to a phantom SnapshotOffer.
In all the BAD cases, these two lines are immediately followed by the Snapshot offer
```
akka.persistence.journal.postgresql] Started (Akka.Persistence.PostgreSql.Journal.PostgreSqlJournal)
akka.persistence.snapshot-store.postgresql] Started (Akka.Persistence.PostgreSql.Snapshot.PostgreSqlSnapshotStore)
```
and only after THAT, the ReplayFilter creation is logged:
```
[akka://TitleCatalogizerPPTEST/system/akka.persistence.journal.postgresql/$a] Started (Akka.Persistence.Journal.ReplayFilter)
[akka://TitleCatalogizerPPTEST/system/akka.persistence.journal.postgresql] now supervising akka://TitleCatalogizerPPTEST/system/akka.persistence.journal.postgresql/$a
```
in all the GOOD cases, the order is the other way around:
```
[akka://TitleCatalogizerPPTEST/system/akka.persistence.journal.postgresql] Started (Akka.Persistence.PostgreSql.Journal.PostgreSqlJournal)
[akka://TitleCatalogizerPPTEST/system/akka.persistence.snapshot-store.postgresql] Started (Akka.Persistence.PostgreSql.Snapshot.PostgreSqlSnapshotStore)

[akka://TitleCatalogizerPPTEST/system/akka.persistence.journal.postgresql/$a] Started (Akka.Persistence.Journal.ReplayFilter)
[akka://TitleCatalogizerPPTEST/system/akka.persistence.journal.postgresql] now supervising akka://TitleCatalogizerPPTEST/system/akka.persistence.journal.postgresql/$a

[[akka://TitleCatalogizerPPTEST/user/TitleCatalogizer/Queue/RuntimeActor#384178050]] Recovering an expected Divv.XiXi.TitleCatalogizer.Actors.Runtime.RuntimeActor+EnqueuedWork
```
I've just done this 15 times, and it's true for all 7 failing and all 8 OK cases.

So there seems to be some kind of race condition, and maybe this gives an

Tagging @Horusiath for his knowledge of Persistence.

**Comment by boekabart** [comment]
Updates:
* Problem is reproducible 100% when using the inmem Journal
* Problem never happens when using the no-snapshot-store

So it seems the bug is isolated in the Snapshot code only.

**Comment by boekabart** [comment]
Used Wireshark to see what Query is executed:
```
                SELECT persistence_id,
                    sequence_nr, 
                    created_at, 
                    manifest, 
                    payload,
                    serializer_id
                FROM public.event_journal 
                WHERE persistence_id = $1 
                    AND sequence_nr <= $2
                    AND created_at <= $3
                ORDER BY sequence_nr DESC
                LIMIT 1
```
It's obviously the correct query (for the most recent snapshot), only the table name is wrong.

**Comment by Horusiath** [comment]
Is #67 possibly related to this issue?

**Comment by basbossink** [comment]
Well.... The pull request #67 does claim to solve this issue 😁😇. And yes, I'm a colleague of @boekabart and he informed of this issue.

**Comment by Arkatufus** [comment]
Closing this issue as resolved since #67 is merged.
